### PR TITLE
ethereum-types upgrade

### PR DESCRIPTION
**Pull request by mdben1247**
#249
Upgraded `ethereum-types` and other stuff.
```
ethereum-types -> 0.9.2
rlp -> 0.4.6
keccak-hash -> 0.5.0
parity-crypto -> 0.6.2
ethabi -> 12.0.0
ethabi-derive -> 12.0.0
ethabi-contract -> 11.0.0
ethbloom -> 0.9.1
rand -> 0.7.3
trie-standardmap -> 0.15.2
triehash -> 0.5.0
```

### CODE REVIEW

**Comment by mdben1247** [comment]
I managed to fix the POA tests and an `ec_recover` issue resposible for InvalidStateRoot on some blocks. I am still fixing other (mostly insignificant thus far) test issues. It's a bit slow going since every time I fix the current batch, a new one opens... I am using `cargo test --all`, is there some better way to do this, that would immediately return all test failures?

**Comment by mdben1247** [comment]
This feels solid now. Live mainnet version is working well and all tests complete successfully, except a doc-test for `TypedTransaction`, but that's been failing on `dev` base too.

**Comment by rakita** [comment]
Yea, we just recently noticed that tests are failing in `dev`. It will be fixed here: https://github.com/openethereum/openethereum/pull/255

**Comment by rakita** [comment]
I will check resolve conflicts.

**Comment by mdben1247** [comment]
In a separate branch I have also upgraed `kvdb-*` and related libraries, which finally brings performance on par with `v3.0.1`. Not sure though, should I push those changes here as a one big package together with `ethereum-types` upgrades or make a separate PR for that?

**Comment by rakita** [comment]
Again, glad to hear this, I already started to review the current pull request and it is a little bit easier to review because changes are what we expect. For crypto, i will need to double-check them but it seems okay. But for database lib upgrades i would like to do some performance testing before we merge it just to be okay that spikes in disk consumption that we noticed in 2.7.x are not tied with upgraded lib or some miss-configuration. It would be good to leave them as separate branches.

**Review by rakita — COMMENTED** [review]
the first iteration done, it seems 95% is okay, most of my comments are for bookmarking porpose. I Will reiterate again and make changes.

**Comment by mdben1247** [comment]
When upgrading, whenever I was in doubt, I looked into `3.0.1` for guidance and mostly copied or adapted from that.

File: crates/accounts/ethkey/Cargo.toml [hunk 1]
```diff
@@ -6,15 +6,15 @@ authors = ["Parity Technologies <admin@parity.io>"]
 
 [dependencies]
 edit-distance = "2.0"
-parity-crypto = "0.3.0"
-eth-secp256k1 = { git = "https://github.com/paritytech/rust-secp256k1", rev = "ccc06e7480148b723eb44ac56cf4d20eec380b6f" }
-ethereum-types = "0.4"
+parity-crypto = { version = "0.6.2", features = ["publickey"] }
+eth-secp256k1 = { git = "https://github.com/paritytech/rust-secp256k1", rev = "9791e79f21a5309dcb6e0bd254b1ef88fca2f1f4" }
```
[the remark this answers is not in the document]
Comment (mdben1247) [reply crates/accounts/ethkey/Cargo.toml:10]: I think this is from before I moved everything to `parity-common` packages, the version before the upgrade had crypto in tree. Anyway, this dependency is not needed anymore and can be removed.

File: crates/concensus/ethash/src/cache.rs [hunk 2]
```diff
@@ -318,27 +318,28 @@ impl AsRef<[Node]> for NodeCache {
 // out. It counts as a read and causes all writes afterwards to be elided. Yes, really. I know, I
 // want to refactor this to use less `unsafe` as much as the next rustacean.
 unsafe fn initialize_memory(memory: *mut Node, num_nodes: usize, ident: &H256) {
-    let dst = memory as *mut u8;
+    // We use raw pointers here, see above
+    let dst = slice::from_raw_parts_mut(memory as *mut u8, NODE_BYTES);
 
     debug_assert_eq!(ident.len(), 32);
-    keccak_512::unchecked(dst, NODE_BYTES, ident.as_ptr(), ident.len());
+    keccak_512::write(&ident[..], dst);
```
[the remark this answers is not in the document]
Comment (rakita) [reply crates/concensus/ethash/src/cache.rs:325]: It is same thing, but a little bit nicer

File: crates/concensus/ethash/src/compute.rs [hunk 3]
```diff
@@ -146,23 +146,18 @@ pub fn quick_get_difficulty(
...
-            keccak_512::unchecked(buf.as_mut_ptr(), 64, buf.as_ptr(), 40);
-            ptr::copy_nonoverlapping(mix_hash.as_ptr(), buf[64..].as_mut_ptr(), 32);
+            keccak_512::inplace_range(&mut buf, 0..end);
+            buf[64..].copy_from_slice(mix_hash);
 
-            // This is initialized in `keccak_256`
-            let mut hash: [u8; 32] = ::mem::MaybeUninit::uninit().assume_init();
-            keccak_256::unchecked(hash.as_mut_ptr(), hash.len(), buf.as_ptr(), buf.len());
+            let mut hash = [0u8; 32];
+            keccak_256::write(&buf, &mut hash);
```
[the remark this answers is not in the document]
Comment (rakita) [reply crates/concensus/ethash/src/compute.rs:160]: seems okay, will try to move unsafe to only cover mem::transmute

File: crates/concensus/miner/src/pool/mod.rs [hunk 4]
```diff
@@ -177,7 +177,7 @@ impl txpool::VerifiedTransaction for VerifiedTransaction {
     }
 
     fn mem_usage(&self) -> usize {
-        self.transaction.heap_size_of_children()
+        self.transaction.malloc_size_of()
```
[the remark this answers is not in the document]
Comment (rakita) [reply crates/concensus/miner/src/pool/mod.rs:180]: HeapSize used by servo was disconniuted and malloc became internal servo lib that parity guys extracted to parity-mem lib

File: crates/ethcore/blockchain/src/blockchain.rs [hunk 5]
```diff
@@ -1817,12 +1817,12 @@ impl BlockChain {
 
     /// Get current cache size.
     pub fn cache_size(&self) -> CacheSize {
+        let mut ops = new_malloc_size_ops();
```
[the remark this answers is not in the document]
Comment (rakita) [reply crates/ethcore/blockchain/src/blockchain.rs:1820]: Okay HeapSize is deprecated lib build by serde team, MallocSize is their current internal lib.

**Comment by adria0** [comment]
Superseded by https://github.com/openethereum/openethereum/pull/315

**Pull request closed without merging**